### PR TITLE
chore: refactoring deploys

### PR DESCRIPTION
**Pull request by smashercosmo**
**Summary**

Deploys refactoring:
1. Covert to TS
2. Proper deploys action type
3. Replace immutable with immer

**Test plan**

Run current tests

### CODE REVIEW

File: packages/netlify-cms-core/src/reducers/deploys.js [hunk 1]
```diff
@@ -1,46 +0,0 @@
...
-      );
-    }
-
-    default:
-      return state;
-  }
-}
-
-export function selectDeployPreview(state, collection, slug) {
-  return state.getIn(['deploys', `${collection}.${slug}`]);
```
Comment (smashercosmo) [inline packages/netlify-cms-core/src/reducers/deploys.js:43]: I removed extra nesting (it was state.deploys.deploys before)

File: packages/netlify-cms-core/src/actions/deploys.js [hunk 2]
```diff
@@ -1,86 +0,0 @@
...
-/**
- * Requests a deploy preview object from the registered backend.
- */
-export function loadDeployPreview(collection, slug, entry, published, opts) {
-  return async (dispatch, getState) => {
-    const state = getState();
-    const backend = currentBackend(state.config);
-
-    // Exit if currently fetching
-    const deployState = selectDeployPreview(state, collection, slug);
```
Comment (smashercosmo) [inline packages/netlify-cms-core/src/actions/deploys.js:52]: I pretty sure this was a bug. selectDeployPreview function doesn't accept collection object.

**Comment by smashercosmo** [comment]
e2e tests are a bit flaky. sometimes they pass, sometimes they don't.

**Comment by erezrokah** [comment]
> e2e tests are a bit flaky. sometimes they pass, sometimes they don't.

Yes I noticed that. I'm still working on that - see failed attempt here https://github.com/netlify/netlify-cms/pull/5131#issue-594928546

File: packages/netlify-cms-core/src/actions/deploys.js (hunk 2 above)
Comment (erezrokah) [reply packages/netlify-cms-core/src/actions/deploys.js:52]: Definitely a bug coming from inconsistent naming (`collection` vs `collectionName`).
Not the first time I've seen it happen in the codebase. I'm glad you fixed it

File: packages/netlify-cms-core/src/reducers/deploys.ts [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+};
+
+const defaultState: Deploys = {};
+
+const deploys = produce((state: Deploys, action: DeploysAction) => {
+  switch (action.type) {
+    case DEPLOY_PREVIEW_REQUEST: {
+      const { collection, slug } = action.payload;
+      state[`${collection}.${slug}`] = {
+        ...state[`${collection}.${slug}`],
```
Comment (erezrokah) [inline packages/netlify-cms-core/src/reducers/deploys.ts:24]: This is required so when clicking the `Check for preview" button, it doesn't disappear

File: packages/netlify-cms-core/src/reducers/deploys.ts [hunk 4]
```diff
@@ -21,6 +21,7 @@ const deploys = produce((state: Deploys, action: DeploysAction) => {
     case DEPLOY_PREVIEW_REQUEST: {
       const { collection, slug } = action.payload;
       state[`${collection}.${slug}`] = {
+        ...state[`${collection}.${slug}`],
```
Comment (smashercosmo) [inline packages/netlify-cms-core/src/reducers/deploys.ts:24]: @erezrokah I would rather do it like this (a bit more explicit)
```js
const key = `${collection}.${slug}`
state[key] = state[key] || {}
state[key].isFetching = true
```

**Pull request merged**